algorithms/mopt/mopt_hierarflow_routines.cpp

```diff
@@ -0,0 +1,213 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2022 Ricerca Security
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+#include "fuzzuf/algorithms/mopt/mopt_hierarflow_routines.hpp"
+
+#include "fuzzuf/algorithms/mopt/mopt_optimizer.hpp"
+#include "fuzzuf/algorithms/mopt/mopt_option.hpp"
+#include "fuzzuf/algorithms/mopt/mopt_option_get_splice_cycles.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+namespace fuzzuf::algorithm::mopt::routine {
+
+namespace other {
+
+using mopt::option::MOptTag;
+
+MOptUpdate::MOptUpdate(MOptState &state) : state(state) {}
+
+MOptMidCalleeRef MOptUpdate::operator()(
+    [[maybe_unused]] std::shared_ptr<MOptTestcase> testcase) {
+  // retrive optimizer values
+  auto last_splice_cycle = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::LastSpliceCycle, true);
+  auto havoc_operator_finds = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::HavocOperatorFinds, true);
+
+  auto &mopt = state.mopt;
+
+  // update havoc_operator_finds
+  for (size_t i = 0; i < havoc_operator_finds.size(); i++) {
+    mopt->havoc_operator_finds[state.core_mode ? 1 : 0][i] +=
```

Comment:
This may be a minute thing, but it is a bit unclear what the immediate values 0 and 1 are pointing out.
I would prefer something like `havoc_operator_finds[state.core_mode ? CORE_MODE : PILOT_MODE]` or  `havoc_operator_finds[state.core_mode]` (by making state.core_mode `enum MOptMode`).

Reply:
fixed in 896319460a5f1cd0bb1c0934868b610193412ccf

Reply:
awesome

Reply:
Oh wait there seem some parts you forgot to fix.

https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-77ad669e643808933880a044def0763a1788478a9f9a42bea615b130fdb42438R79-R87
https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-77ad669e643808933880a044def0763a1788478a9f9a42bea615b130fdb42438R99-R102

Reply:
https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-6a8c4972fbfb3703282f4f41c9b8a685401f29bc8c395d701a71bb15fdf0d496R58-R61

Reply:
https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-6a8c4972fbfb3703282f4f41c9b8a685401f29bc8c395d701a71bb15fdf0d496R118

Reply:
fixed in 28bc3f78cad5b18ff8050e71fbad85363d630343

Reply:
Overlooking one:
https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-6a8c4972fbfb3703282f4f41c9b8a685401f29bc8c395d701a71bb15fdf0d496R116

Reply:
I intentionally left this one as it doesn't have nothing to do with the type of MOptMode.

Reply:
Then why not use `NUM_CASE` ? Are those sometimes not the same?

Reply:
the point makes sense. applied in 48ca48478d07834d188c4590f0f692a0dec6aee2